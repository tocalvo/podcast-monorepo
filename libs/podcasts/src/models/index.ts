export * from './podcasts.model';
export * from './podcastDetail.model';
