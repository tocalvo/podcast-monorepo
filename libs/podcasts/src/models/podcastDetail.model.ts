// Talk with backend to know their structure/swagger and follow the same conventions, or autogenerate models
// https://github.com/openapi-ts/openapi-typescript/tree/main/packages/openapi-typescript
export interface ITunesPodcastDetailQuery {
  resultCount: number;
  results: ITunesPodcastDetail[];
}

export enum ITunesPodcastKind {
  podcast = 'podcast',
  'podcast-episode' = 'podcast-episode',
}

export interface ITunesPodcastDetail {
  artistIds?: number[];
  artistId?: number;
  artistName?: string;
  artistViewUrl: string;
  artworkUrl30?: string;
  artworkUrl60?: string;
  artworkUrl100?: string;
  artworkUrl600?: string;
  collectionCensoredName?: string;
  collectionExplicitness?: string;
  collectionHdPrice?: number;
  collectionId: number;
  collectionName: string;
  collectionPrice?: number;
  collectionViewUrl: string;
  contentAdvisoryRating: string;
  country: string;
  currency?: string;
  feedUrl: string;
  genreIds?: string[];
  genres: string[] | { name: string; id: string }[];
  shortDescription?: string;
  description?: string;
  episodeUrl?: string;
  episodeFileExtension?: string;
  kind: ITunesPodcastKind;
  primaryGenreName: string;
  releaseDate: string;
  trackCensoredName: string;
  trackCount: number;
  trackExplicitness: string;
  trackId: number;
  trackName: string;
  trackPrice: number;
  trackTimeMillis: number;
  trackViewUrl: string;
  wrapperType: string;
}

// extended to add calculated data
export interface ITunesPodcastDetailQueryExtended
  extends ITunesPodcastDetailQuery {
  results: ITunesPodcastDetailExtended[];
}

export interface ITunesPodcastDetailExtended extends ITunesPodcastDetail {
  releaseDateLocaleString: string;
  trackTimeMinsSegs: string;
  sanitizedDescription: string;
}
