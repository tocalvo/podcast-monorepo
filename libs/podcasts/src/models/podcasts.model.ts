// Talk with backend to know their structure/swagger and follow the same conventions, or autogenerate models
// https://github.com/openapi-ts/openapi-typescript/tree/main/packages/openapi-typescript

export interface ITunesPodcastLabel {
  label: string;
}
export interface ITunesPodcastLink {
  attributes: {
    href: string;
    rel: string;
    type?: string;
  };
}

export interface ITunesPodcastEntry {
  category: {
    attributes: {
      'im:id': string;
      label: string;
      scheme: string;
      term: string;
    };
  };
  id: {
    attributes: {
      'im:id': string;
    };
    label: string;
  };
  'im:artist': {
    attributes: {
      href: string;
    };
    label: string;
  };
  'im:contentType': {
    attributes: {
      term: string;
      label: string;
    };
  };
  'im:image': {
    attributes: {
      height: string;
    };
    label: string;
  }[];
  'im:name': ITunesPodcastLabel;
  'im:price': {
    attributes: {
      amount: string;
      currency: string;
    };
  };
  'im:releaseDate': {
    attributes: ITunesPodcastLabel;
    label: string;
  };
  link: ITunesPodcastLink;
  rights: ITunesPodcastLabel;
  summary: ITunesPodcastLabel;
  title: ITunesPodcastLabel;
}

export interface ITunesPodcasts {
  feed: {
    author: {
      name: ITunesPodcastLabel;
      uri: ITunesPodcastLabel;
    };
    entry: ITunesPodcastEntry[];
    icon: ITunesPodcastLabel;
    id: ITunesPodcastLabel;
    link: ITunesPodcastLink[];
    rights: ITunesPodcastLabel;
    title: ITunesPodcastLabel;
    updated: ITunesPodcastLabel;
  };
}
