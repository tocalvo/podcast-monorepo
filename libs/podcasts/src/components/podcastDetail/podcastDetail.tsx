import { useMemo } from 'react';
import { useItunesPodcastDetail, useItunesPodcasts } from '../../hooks';
import styles from './podcastDetail.module.css';
import { ITunesPodcastKind } from '../../models';
import PodcastBar from '../podcastBar/podcastBar';
import { Link, useParams } from 'react-router';

export const PodcastDetail: React.FC = () => {
  const { podcastId } = useParams();

  const itunesPodcastsQuery = useItunesPodcasts();
  const itunesPodcast = useMemo(() => {
    return itunesPodcastsQuery.data?.feed.entry.find(
      (podcast) => podcast.id.attributes['im:id'] === podcastId
    );
  }, [itunesPodcastsQuery.data, podcastId]);

  const podcastDetailQuery = useItunesPodcastDetail(podcastId);
  const podcastEpisodes = useMemo(() => {
    return (
      podcastDetailQuery.data?.results.filter(
        (podcast) => podcast.kind === ITunesPodcastKind['podcast-episode']
      ) || []
    );
  }, [podcastDetailQuery.data]);

  if (!itunesPodcast) {
    return; // redirect to home or to 404
  }

  return (
    <div className={styles['container']}>
      <PodcastBar podcast={itunesPodcast} />
      <div className={styles['episodes-container']}>
        <div className={styles['card']}>
          <p>
            <b>Episodes: {podcastEpisodes?.length}</b>
          </p>
        </div>
        <div className={`${styles['card']} ${styles['striped-table']}`}>
          <table>
            <thead>
              <tr>
                <th>Title</th>
                <th>Date</th>
                <th>Duration</th>
              </tr>
            </thead>
            <tbody>
              {podcastEpisodes.map((episode, i) => (
                <tr key={i}>
                  <td>
                    <Link
                      className={styles['link']}
                      to={`/podcast/${podcastId}/episode/${episode.trackId}`}
                    >
                      {episode.trackName}
                    </Link>
                  </td>
                  <td>{episode.releaseDateLocaleString}</td>
                  <td>{episode.trackTimeMinsSegs}</td>
                </tr>
              ))}
            </tbody>
          </table>
        </div>
      </div>
    </div>
  );
};
export default PodcastDetail;
