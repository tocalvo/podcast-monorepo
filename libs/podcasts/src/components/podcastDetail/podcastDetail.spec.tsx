import { render, screen, within } from '@testing-library/react';

import { PodcastDetail } from './podcastDetail';
import { podcastListMock } from '../../__mocks__/podcast.mock';
import { podcastDetailMock } from '../../__mocks__/podcastDetail.mock';

import { BrowserRouter } from 'react-router';
import { ITunesPodcastKind } from '../../models';

vi.mock('../../hooks', () => ({
  useItunesPodcasts: vi.fn(() => ({
    podcasts: podcastListMock.feed.entry,
    getPodcastResumeById: vi
      .fn()
      .mockReturnValue(podcastListMock.feed.entry[0]),
  })),
  useItunesPodcastDetail: vi.fn(() => ({
    podcastDetail: podcastDetailMock,
    episodes: podcastDetailMock.results.filter(
      (podcast) => podcast.kind === ITunesPodcastKind['podcast-episode']
    ),
    findEpisodeById: vi.fn().mockReturnValue(podcastDetailMock.results[1]),
  })),
}));

vi.mock('react-router', async () => {
  const actual = await vi.importActual('react-router');
  return {
    ...actual,
    useParams: vi.fn(() => ({
      podcastId: podcastListMock.feed.entry[0].id.attributes['im:id'],
    })),
  };
});

describe('PodcastDetail', () => {
  it('should render successfully', () => {
    const { baseElement, container } = render(
      <BrowserRouter>
        <PodcastDetail />
      </BrowserRouter>
    );
    expect(baseElement).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it('should render the podcasts bar', () => {
    render(
      <BrowserRouter>
        <PodcastDetail />
      </BrowserRouter>
    );

    expect(screen.queryByTestId('podcasts_bar')).toBeTruthy();
  });

  it('should render the episodes from the mock', () => {
    render(
      <BrowserRouter>
        <PodcastDetail />
      </BrowserRouter>
    );
    const rows = screen.getAllByRole('row');
    expect(rows.length).toBe(podcastDetailMock.results.length);
    // check titles
    const { getAllByRole } = within(rows[0]);
    const cells = getAllByRole('columnheader');

    expect(cells.length).toBe(3);
    expect(cells[0].textContent).toBe('Title');
    expect(cells[1].textContent).toBe('Date');
    expect(cells[2].textContent).toBe('Duration');

    // check content
    const { getAllByRole: getAllByRole2 } = within(rows[1]);
    const contentCells = getAllByRole2('cell');

    expect(contentCells.length).toBe(3);
    expect(contentCells[0].querySelector('a')?.getAttribute('href')).toBe(
      '/podcast/1535809341/episode/1000677977070'
    );
    expect(contentCells[0].textContent).toBe(
      'Episode 778 | "Bottom Of The 9th"'
    );
    expect(contentCells[1].textContent).toBe('08/07/1987');
    expect(contentCells[2].textContent).toBe('12:00');
  });
});
