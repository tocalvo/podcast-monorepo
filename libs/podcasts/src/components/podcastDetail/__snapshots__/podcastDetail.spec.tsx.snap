// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PodcastDetail > should render successfully 1`] = `
<div>
  <div
    class="_container_10c5fc"
  >
    <div
      class="_card_33fd63"
      data-testid="podcasts_bar"
    >
      <a
        class="_image-link_33fd63"
        data-discover="true"
        href="/podcast/1535809341"
      >
        <img
          alt=""
          class="_image_33fd63"
          src="https://is1-ssl.mzstatic.com/image/thumb/Podcasts113/v4/f2/21/fa/f221fabd-017f-5125-633b-f1fe4f39802a/mza_182995249085044287.jpg/55x55bb.png"
        />
      </a>
      <hr
        class="_line_33fd63"
      />
      <a
        data-discover="true"
        href="/podcast/1535809341"
      >
        <p
          class="_title_33fd63"
        >
          The Joe Budden Podcast
        </p>
      </a>
      <a
        data-discover="true"
        href="/podcast/1535809341"
      >
        <p
          class="_italic_33fd63"
        >
          by 
          The Joe Budden Network
        </p>
      </a>
      <hr
        class="_line_33fd63"
      />
      <p
        class="_title2_33fd63"
      >
        Description:
      </p>
      <p
        class="_italic_33fd63"
      >
        Tune into Joe Budden and his friends. Follow along the crazy adventures of these very random friends.
      </p>
    </div>
    <div
      class="_episodes-container_10c5fc"
    >
      <div
        class="_card_10c5fc"
      >
        <p>
          <b>
            Episodes: 
            2
          </b>
        </p>
      </div>
      <div
        class="_card_10c5fc _striped-table_10c5fc"
      >
        <table>
          <thead>
            <tr>
              <th>
                Title
              </th>
              <th>
                Date
              </th>
              <th>
                Duration
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                <a
                  class="_link_10c5fc"
                  data-discover="true"
                  href="/podcast/1535809341/episode/1000677977070"
                >
                  Episode 778 | "Bottom Of The 9th"
                </a>
              </td>
              <td>
                08/07/1987
              </td>
              <td>
                12:00
              </td>
            </tr>
            <tr>
              <td>
                <a
                  class="_link_10c5fc"
                  data-discover="true"
                  href="/podcast/1535809341/episode/1000677608756"
                >
                  Episode 777 | "The Spectrum"
                </a>
              </td>
              <td>
                08/07/1987
              </td>
              <td>
                12:00
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
